### ISSUE TITLE
Image not visible, if parent Sprite has not been added to stage - starling 2.0

### ISSUE THREAD

**Issue by syberkitten** [Gamua/Starling-Framework#821]
Just made a migration to Starling 2.0.

all seems to work except this strange bug:
adding Image instance to a Sprite before the latter is added to the stage
does not render the Image.

My workaround is to apply a filter on each image, and then set it back
to null, to get the Images showing.

In another case, where a parent
Sprite has already been added to the starling stage (before the Images are added to the parent),
it works fine, and images are rendered as they were in version 1.x ...

Is there any difference in starling 2.0 as to bitmapData is loaded into
the Texture for the Image?

I'm extending the Image class and my constructor creates a custom texture:
`var t:Texture = Texture.fromBitmapData(myBitmapData,false,false,1);
super(t);
`

Thanks

**Comment by PrimaryFeather** [comment]
Could you send me a complete sample showcasing this problem? I haven't heard of such issues before, and I'm quite sure I added an image to a sprite before adding it to the stage, without problems. We must be looking at a side effect of something you're doing, and I'm not. :wink:

Thanks a lot in advance!

**Comment by syberkitten** [comment]
Hi Daniel, Now you have me challenged ;)
I'm also sure it has to do with some side effects, as if this
issue was straight forward, almost all applications would probably fail.

it will be somewhat not simple to reproduce since our system
loads jpg files from the android file system and creates Images
instance that are added to the starling Stage one by one, as they are loaded.

for each jpg image:
- a sprite is created
- image instance is added into the sprite, after it's Texture is created.
- sprite receives some transformation
- sprite is added to stage


In version 1.x there were no problems and Images were shown immediately,
yet in 2.0 they are as if invisible, but actually reside on the stage and react to gesture events.

every tap on a sprite highlights the Image for a second,
thats when they become visible again, when the first filter is added.

could you give me some possible things to check, that come to mind?

Thanks -))

**Comment by syberkitten** [comment]
Wanting to add, that these "invisible" Images start to become "visible" 
either by applying a filter, or by being moved... then they are simply "show up"

**Comment by PrimaryFeather** [comment]
Hm ... do they also show up if you call `image.setRequiresRedraw()` or `sprite.setRequiresRedraw()`? That's what tells the render cache that the objects have changed.

You say you're loading JPG images, right? Not ATF? I.e. the textures are loaded synchronously.

**Comment by syberkitten** [comment]
Ok, this is a good lead -) and this is where it got me:

when using `image.setRequiresRedraw();` after a certain interval, the Image is shown, same was when i applied the filter as a workaround to show the Image (it was also after a certain interval time).

which most likely means, the problem is the loop that creates the Image instances.

before version 2.0 once all Images were added to stage they would appear all at once
since the loop is 'blocking' and they would all be ready on the same frame.

In starling 2.0 it looks like such style of creating Image instances somehow does not render them,
only after a change of state after the next frame was rendered.

What is the change in the internals of Starling 2.0 vs. 1.x in this regard?
ps. I'm using AIR Sdk 20

**Comment by PrimaryFeather** [comment]
<blockquote>What is the change in the internals of Starling 2.0 vs. 1.x in this regard?</blockquote> 

Starling now tries to render unchanged parts of the display list directly from a new cache, which can have great performance / battery benefits. But it has some side effects. :wink:

Normally, whenever a crucial change is done to an image (say, it's position, color, or texture changes), the method `setRequiresRedraw()` will be called automatically, which will force the object to be drawn directly (not from cache) in the next frame. 

With textures, this is sometimes not possible, because the texture itself doesn't have a reference to the images it's used on, so it can't tell them to redraw. My guess is that we're looking at such a situation.

I know it's time consuming, but could you try to create a demo project that loads the images just as you describe? I'm sure when I've got a sample that I can use to reproduce the problem, it's easily fixed. (Famous last words, right?)

**Comment by PrimaryFeather** [comment]
Were you able to reproduce this in some way?

**Comment by PrimaryFeather** [comment]
Did this pop up again, or maybe go away with the release version of Starling 2?

**Comment by syberkitten** [comment]
Hi Daniel
thanks for the follow up here :)

I did not Isolate the issue yet.
But did try it with Starling 2 latest dev branch a week ago, and it did not go away,
i even had other side effects that were not expected,
so i just reverted back before needing to catch my head with both hands!

my scenario is that I'm simply adding images from a file system unto the stage, each image's bitmapdata
is loaded as a texture into an Image instance and added to stage.
Until the images are transformed after they are added, they are not show.

I'm using Gestouch library to add touch events to these Images
and I can select an invisible image and start dragging it, then it
magically appears on Stage.

the scenario is that these images are added in one iteration loop with callbacks,
once an image's data is ready, it's added, so we can have between one and x (mostly between 1-20)
images that are loaded one by one.

Images actual pictures and not textures, so they are heavier and depend mostly on the device's camera.
hopefully soon, I'll have more time to play with Starling 2 again and try different approaches.

If all fails again, i will try to create a demo app like you suggested. 👍

**Comment by PrimaryFeather** [comment]
Thanks a lot for the detailed response, Liam!
Mhm, I understand that you reverted to v1.8 — so far into development, that's definitely smart; it's a very solid release and will work well for you!

It must have something to do with the texture, but I don't see yet how there could be problem with JPG textures (ATF would be another story). 

In any case, if you ever come around to create that demo app, that would be a huge help for me! Just let me know — otherwise, we'll just have to wait until someone else runs into the same issue; perhaps in an easier-to-reproduce scenario. 

Cheers!

**Comment by fraylopez** [comment]
Hi,

I'm also having such behaviour or at least something similar. It looks like it is something related to nested LayoutGroups. I will try to isolate the issue into a test project.

This started happening after 2.01 migration on Android  with AIR 22.0 (Win is OK)

On the attached pic, the textures not showing at the upper bar are icons for Feather Buttons arranged into an horizontal layout.

![error_textures](https://cloud.githubusercontent.com/assets/12448436/18266175/af98f40c-7419-11e6-9923-faf4d0fba333.png)


Thanks!

**Comment by PrimaryFeather** [comment]
To whoever is having similar problems: I recently changed some of the render-cache logic, in the hope to make it more reliable. Please retry with the latest head revision, and tell me if that helps!

Thanks a lot in advance! :smile:

**Comment by syberkitten** [comment]
Hi
I've tried the v2.1, but did not observe a significant difference towards these issues: 

- Images are not drawn, but only when added to stage with a delay, and using "setRequiresRedraw" on the parent (containing Sprite)
- When adding a Blur filter as a FragmentFilter, Image becomes too slow to move around (with touch gestures)
when trying to cache() the FragmentFilter or the FilterChain instance, the image is simply not visible (not drawn at all).

A simple work I plan to use is create a blur effect
whether classic DisplayList or a FragmentFilter, and reflect it on a Texture to be loaded unto an Image instance; 
which is perhaps a reasonable practice to use filters anyway.. ?
given that they take such a toll on the GPU.

ps. when moving around an Image with a BlurFilter with touch events, the frame rate
drops to around 2-3, (slower in my testing then the classic DisplayList equivalent)

Thanks a lot for following up on this issue 👍 :)

**Comment by PrimaryFeather** [comment]
It's a bummer that my changes did not fix the problem — but it was worth a shot, in any case.

Again, do you think it would be possible to try and reproduce this problem in a separate project? Or to strip down the project extremely, to have something I can try out? I think on the long run, that's the only way I'm going to be able to get to the root of this.

As for the blur filter performance: how much blur are you using / what are the values of `blurX` and `blurY`? Because blur can generate quite a lot of draw calls (e.g. `blurX = blurY = 4` would cause 8 draw calls).

**Comment by PrimaryFeather** [comment]
Since I could not yet reproduce this issue, I'll close it for now — feel free to open it up again if this is still happening.

As for the _BlurFilter_, Starling 2.2 greatly enhanced the blurriness of the filter, i.e. you get a higher blur with fewer draw calls. So at least that part should be truly solved. :smile:
